### ISSUE TITLE
implementing del keyword to delete objects

### ISSUE THREAD

**Issue by rakati** [pyccel/pyccel#1207]
**Describe the feature**
we already have our small garbage collector for the memory we allocate during pyccelizing python code, but we don't have the support for the `del` keyword.

**Test Code**
```python
import numpy as np

mat = np.arange(13)
# doing some calculations using the array mat
del mat
```
we need to free the memory allocated for the array mat.
**Aditional information**
the `del` keyword can be used for any object in python, we can't do that for all types (int, float...) because pyccel doesn't; allocate memory for all of these types, so using it only for the types we allocate memory for, and display warning when the object not allocated.
 
**Language**
Fortran, C

**Comment by yassine-alaoui** [comment]
Hello, @rakati @EmilyBourne, I tried to see if I could implement this function in C and Fortran and found that it already existed in Fortran, but had some issues and didn't exist in c, and when I tried to add it I found a function called ```free_array()``` already implemented in c and I found out that the code already used this function to free any array that was allocated; once an array gets allocated then the code adds the part of the array being freed at the end of the program, like a garbage collector.  And when I implemented this ```Del``` function using the same function I got something like this:
```c
#include "test.h"
#include <stdlib.h>
#include "ndarrays.h"
#include <math.h>
#include <stdint.h>
int main()
{
    t_ndarray mat = {.shape = NULL};
    int64_t i;
    mat = array_create(1, (int64_t[]){(int64_t)ceil((double)((INT64_C(13) - INT64_C(0))) / 1.0)}, nd_int64, false);
    for (i = INT64_C(0); i < mat.shape[INT64_C(0)]; i += INT64_C(1))
    {
        GET_ELEMENT(mat, nd_int64, (int64_t)i) = i;
    }
    /*doing some calculations using the array mat*/
    free_array(mat);
    free_array(mat);
    return 0;
}
```
out of this code bracket:
```Python
import numpy as np
if "__name__" == "__main__":
    mat = np.arange(13)
    # doing some calculations using the array mat
    del mat
```
The duplicated free_array doesn't show any errors in this case.
If we remove ```del mat``` however then we don't get any duplicates; like this :
```c
.
.
.
    }
    /*doing some calculations using the array mat*/
    free_array(mat);
    return 0;
}
```

**Comment by EmilyBourne** [comment]
@yassine-alaoui In my opinion the duplication is ok. As long as we are setting pointers to `NULL` there shouldn't be any errors. It is simpler for us to free all memory at the end than to try and work out what is already freed (what happens if it is freed in one half of an if/else block).

The code for Fortran is nearly as old as pyccel so it doesn't surprise me if it doesn't work well. If you decide to tackle this low-priority issue then don't forget to ensure the allocation status is correctly set if we re-allocate after calling `del`

**Comment by EmilyBourne** [comment]
Fixed by #1508
